### PR TITLE
Strengthen story fingerprinting (#15)

### LINKED ISSUE #15: Strengthen story fingerprinting to reduce duplicate publishes
## Summary
- Make story fingerprints more stable by combining deterministic article metadata with LLM-derived key facts.

## Context
- Current fingerprints are hashes of normalized key facts, with headline fallback. Since key facts are generated text, small wording changes can create a new fingerprint for the same underlying story. That weakens duplicate prevention and update detection across cycles.

## Scope
- In scope:
  - Include deterministic signals such as normalized source URLs, entity IDs, topic/category, and a publication/completion date bucket.
  - Keep fingerprints stable enough for updates while avoiding collisions between distinct stories.
  - Update tests around duplicate prevention and update matching.
- Out of scope:
  - Replacing editorial ranking or clustering wholesale.

## Triage Tag
- Priority: High

## Acceptance Criteria
- [ ] Same story with slightly different LLM key-fact wording resolves to the same fingerprint when deterministic signals match.
- [ ] Distinct stories about the same team/player do not collapse incorrectly.
- [ ] URL-overlap fallback still works as a secondary safety net.
- [ ] Tests cover stable same-story and distinct-story examples.

## Tasks (Optional)
- [ ] Define fingerprint input schema.
- [ ] Update `fingerprint()`/plan recomputation flow.
- [ ] Add regression tests for duplicate prevention.

## SOTA Resolution
- Prefer deterministic identifiers and normalized structured metadata for idempotency keys; use LLM text as supporting evidence, not the sole identity source.

## Notes/Links (Optional)
- Relevant files: `app/editorial/helpers.py`, `app/editorial/workflow.py`, `app/schemas.py`.

### PR DESCRIPTION
**Pull request by BigSlikTobi**
Closes #15.

## Summary
- Replace `sha256(sorted key_facts)` with deterministic identity built from normalized source URLs + entity IDs.
- LLM rewording of `key_facts` no longer produces a fresh fingerprint for the same story.
- Distinct stories about the same player/team keep distinct fingerprints because their URL sets differ.
- URL-overlap fallback in `resolve_existing_article_ids` is unchanged and catches legacy `editorial_state` rows for one rollout window — no backfill needed.

## Changes
- `app/editorial/helpers.py`: new `_normalize_url`, `compute_story_fingerprint(urls, entity_ids)`, private `_fingerprint_from_key_facts` (last-resort fallback). `recompute_cluster_fingerprint` switched to URL-only. `recompute_plan_fingerprints(plan, raw_articles=None)` now layers in entity IDs via the existing `story_id → RawArticle` join pattern.
- `app/editorial/workflow.py`: passes `context.raw_articles` into `recompute_plan_fingerprints`.
- `tests/test_helpers.py`: renamed legacy fingerprint tests to `TestKeyFactsFallback`; added `TestComputeStoryFingerprint` (URL normalization, order independence, entity disambiguation) and `TestRecomputePlanFingerprints` (key-fact drift stability, distinct-story separation, entity layering, headline fallback). Updated `TestRecomputeClusterFingerprint` to assert URL-based hashing — including the issue's exact case ("Bills traded a 4th-round pick" vs "Buffalo dealt a fourth-rounder" → same hash when URL matches).

## Acceptance criteria
- [x] Same story / different LLM wording → same fingerprint
- [x] Distinct stories same team/player → different fingerprints
- [x] URL-overlap fallback still active
- [x] Tests cover stable-same-story and distinct-story examples

## Test plan
- [x] `./venv/bin/pytest tests/test_helpers.py -v` (47 passed)
- [x] `./venv/bin/pytest tests/` (169 passed)
- [x] Live cycle dry-run on real Supabase data (`var/output.json`): fingerprints render as clean 16-hex; orchestrator skip/update decisions look correct; one-cycle `prevented_duplicates: 0` quirk is the documented rollout effect (legacy stored fingerprints don't match new scheme; URL-overlap fallback handles it).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/editorial/helpers.py [hunk 1]
```diff
@@ -121,18 +123,45 @@ def group_by_entity(articles: list[RawArticle]) -> GroupedArticles:
...
-    """Compute deterministic fingerprint from all key_facts across source_digests."""
-    all_facts: list[str] = []
-    for digest in cluster.source_digests:
-        all_facts.extend(digest.key_facts)
-    return fingerprint(all_facts)
+    """Compute deterministic fingerprint from normalized source URLs.
+    Falls back to the cluster headline when no URLs are present."""
+    urls = [d.url for d in cluster.source_digests]
+    if urls:
+        return compute_story_fingerprint(urls)
```
Comment (BigSlikTobi) [inline app/editorial/helpers.py]: `recompute_cluster_fingerprint()` is used by `analyze_story_cluster` to set the fingerprint and `is_new` before the orchestrator compares against `published_fingerprints`, but this path now hashes URLs only. Later, `recompute_plan_fingerprints(raw_plan, context.raw_articles)` hashes the same story with entity IDs included. After rollout, published state will contain the URL+entity fingerprint, so an already-published cluster can come back from the tool as `is_new=True` / no matching fingerprint, and the model may choose `publish` or rewrite without update context before post-processing changes the fingerprint again. Can we make the tool-side fingerprint use the same entity-aware input/schema as the final plan recomputation, or avoid changing fingerprints after the action decision?

File: app/editorial/helpers.py [hunk 2]
```diff
@@ -168,24 +197,33 @@ def url_overlap_ratio(
...
+            if ra is None:
+                continue
+            for entity in ra.entities:
+                ent_ids.add(entity.entity_id)
+        if urls:
+            real_fp = compute_story_fingerprint(urls, ent_ids)
         else:
-            # No digests (single-source evaluated by orchestrator) — hash the headline
-            real_fp = fingerprint([story.cluster_headline])
+            real_fp = _fingerprint_from_key_facts([story.cluster_headline])
```
Comment (BigSlikTobi) [inline app/editorial/helpers.py:278]: This fallback means selected `single_source_articles` still do not get URL-based fingerprints. The orchestrator is explicitly told to evaluate singles directly and the output instructions do not require it to populate `source_digests`; when that list is empty we hash the model-generated `cluster_headline`, and persistence later records no `source_urls` for URL-overlap matching. That leaves breaking single-source stories with the same LLM/headline drift problem this PR is fixing for clusters. Can we carry the single-source article id/url into the `StoryEntry` (or synthesize a digest from the compact candidate) so singles also call `compute_story_fingerprint([url], entity_ids)`?

File: app/editorial/helpers.py (hunk 1 above)
Comment (BigSlikTobi) [reply app/editorial/helpers.py]: Good catch — fixed in 9787af6. `recompute_cluster_fingerprint` now takes `entity_ids`, and `tools.py` passes them in from the same `RawArticle` set the plan-level recompute will see. Tool-side fp now equals plan-side fp, so `is_new` is decided against the canonical identity. Added a regression test (`TestRecomputeClusterFingerprint::test_entity_ids_match_plan_recompute`) that locks this invariant in.

File: app/editorial/helpers.py (hunk 2 above)
Comment (BigSlikTobi) [reply app/editorial/helpers.py:278]: Fixed in 9787af6 with two complementary changes:

1. **Prompt**: `editorial_orchestrator_agent` now MUST populate `source_digests` for every single-source `StoryEntry` (one digest carrying the candidate's `id`, `url`, `title`, `source_name`).
2. **Server-side safety net**: new `synthesize_missing_digests(plan, raw_articles)` runs before `recompute_plan_fingerprints` in workflow.py. For any story with empty `source_digests`, it does an exact normalized-title match against `raw_articles` and synthesizes a minimal digest carrying the URL. Ambiguous titles are skipped. Existing digests are never overwritten.

Net effect: single-source stories now get URL-based fingerprints and persist `source_urls`, so the URL-overlap fallback works for them next cycle. Covered by `TestSynthesizeMissingDigests` (5 cases including the end-to-end "synthesis unlocks URL-based fingerprint" path).
